File: src/models/agent_protocol.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Generic
from typing import TypeVar

from pydantic import BaseModel
from pydantic import Field


TData = TypeVar("TData")


class AgentName(str, Enum):
    COST = "cost"
    RESOURCE = "resource"
    HISTORY = "history"
    DIAGNOSIS = "diagnosis"
    REMEDIATION = "remediation"


class AgentStatus(str, Enum):
    OK = "ok"
    DEGRADED = "degraded"
    ERROR = "error"


class AgentResult(BaseModel, Generic[TData]):
    agent: AgentName
    status: AgentStatus
    started_at: datetime
    finished_at: datetime
    data: TData | None = None
    errors: list[str] = Field(default_factory=list)

File: dotnet/src/IncidentWarRoom.Core/Models/AgentProtocol.cs
using System;
using System.Collections.Generic;

namespace IncidentWarRoom.Core.Models;

public enum AgentName
{
    Cost,
    Resource,
    History,
    Diagnosis,
    Remediation,
}

public enum AgentStatus
{
    Ok,
    Degraded,
    Error,
}

public sealed record AgentResult<T>(
    AgentName Agent,
    AgentStatus Status,
    DateTimeOffset StartedAt,
    DateTimeOffset FinishedAt,
    T? Data,
    IReadOnlyList<string> Errors
)
{
    public static AgentResult<T> Ok(AgentName agent, DateTimeOffset startedAt, DateTimeOffset finishedAt, T data) =>
        new(agent, AgentStatus.Ok, startedAt, finishedAt, data, Array.Empty<string>());

    public static AgentResult<T> Degraded(AgentName agent, DateTimeOffset startedAt, DateTimeOffset finishedAt, T? data, params string[] errors) =>
        new(agent, AgentStatus.Degraded, startedAt, finishedAt, data, errors);

    public static AgentResult<T> Error(AgentName agent, DateTimeOffset startedAt, DateTimeOffset finishedAt, params string[] errors) =>
        new(agent, AgentStatus.Error, startedAt, finishedAt, default, errors);
}
